mono/metadata/appdomain.c

```diff
@@ -3312,9 +3312,13 @@ deregister_reflection_info_roots (MonoDomain *domain)
 	mono_domain_assemblies_unlock (domain);
 }
 
+extern MonoCoopMutex mono_domain_unload_mutex;
+
 static gsize WINAPI
 unload_thread_main (void *arg)
 {
+	mono_coop_mutex_lock(&mono_domain_unload_mutex);
```

Comment:
I don't have a strong preference, but we could also put locking in the function which starts/waits on this thread, `mono_domain_try_unload`.